lib/active_merchant/billing/gateways/commerce_hub.rb

```diff
@@ -0,0 +1,327 @@
+          source[:card][:securityCodeIndicator] = 'PROVIDED'
+        end
+      end
+
+      def add_payment_token(source, payment, options)
+        source[:sourceType] = 'PaymentToken'
+        source[:tokenData] = payment
+        source[:tokenSource] = options[:token_source] if options[:token_source]
+        if options[:card_expiration_month] || options[:card_expiration_year]
+          source[:card] = {}
+          source[:card][:expirationMonth] = options[:card_expiration_month] if options[:card_expiration_month]
+          source[:card][:expirationYear] = options[:card_expiration_year] if options[:card_expiration_year]
+        end
+      end
+
+      def add_payment(post, payment, options = {})
+        source = {}
+        if payment.is_a?(NetworkTokenizationCreditCard) && payment.brand == 'google_pay' && options[:google_pay]
+          add_google_pay(source, options)
+        elsif payment.is_a?(NetworkTokenizationCreditCard) && payment.brand == 'apple_pay' && options[:apple_pay]
+          add_apple_pay(source, options)
+        elsif payment.is_a?(CreditCard)
+          add_credit_card(source, payment, options)
+        elsif payment.is_a?(String)
+          add_payment_token(source, payment, options)
+        end
+        post[:source] = source
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def headers(request, options)
+        time = DateTime.now.strftime('%Q').to_s
+        client_request_id = options[:client_request_id] || rand.to_s[2..8]
+        raw_signature = @options[:api_key] + client_request_id.to_s + time + request
+        hmac = OpenSSL::HMAC.digest('sha256', @options[:api_secret], raw_signature)
+        signature = Base64.strict_encode64(hmac.to_s).to_s
+
+        {
+          'Client-Request-Id' => client_request_id,
+          'Api-Key' => @options[:api_key],
+          'Timestamp' => time,
+          'Accept-Language' => 'application/json',
+          'Auth-Token-Type' => 'HMAC',
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json',
+          'Authorization' => signature
+        }
+      end
+
+      def add_merchant_details(post)
+        post[:merchantDetails] = {}
+        post[:merchantDetails][:terminalId] = @options[:terminal_id]
+        post[:merchantDetails][:merchantId] = @options[:merchant_id]
+      end
+
+      def commit(action, parameters, options)
+        url = (test? ? test_url : live_url) + ENDPOINTS[action]
+        add_merchant_details(parameters)
+        response = parse(ssl_post(url, parameters.to_json, headers(parameters.to_json, options)))
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(action, response),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def handle_response(response)
+        case response.code.to_i
+        when 200...300, 400, 401
```

Comment:
I'm getting a bunch of `429` error codes when running the tests. Might be worth adding that to the `when` clause here.

Reply:
I have updated the `handle_response` to include 429 as well. Can you also share the error you are getting because the tests are passing for me except for 3 tests: `test_successful_purchase_with_apple_pay`, `test_successful_purchase_with_google_pay`, and `test_successful_store_with_purchase`